### ISSUE TITLE
Handle HTTP 422 errors from GitHub

### ISSUE THREAD

**Issue by rvl** [bors-ng/bors-ng#1027]
Hi, we are getting "Cancelled" status and crashes when trying to merge a PR. It's a simple PR - single file changed, commit is GPG signed, no squash merges - nothing special.

What has happened is that bors-ng accepted the PR and attempted to merge, but GitHub rejected it due to absence of CODEOWNERS approvals.

I see multiple reports of this issue: #561 #674 #993 #725 #754. And I think @umamaistempo even fixed it once. So this might be a regression.

### Expected behaviour

Bors sends a comment to the PR saying that github rejected the merge with error code 422.

### Actual behaviour

`badmatch` crash dump is only visible in bors-ng history dashboard.

### Example PR

input-output-hk/cardano-node#1823

### Version

a4d04fd740e24d03833cd46a76f0f9e5be96f818 docker image running on our own instance https://bors-ng.aws.iohkdev.io/repositories/13/log#batch-6434

**Comment by umamaistempo** [comment]
:thinking: 

Could you provide the stacktrace for that failure ?

**Comment by rvl** [comment]
Sure... 9/10/2020, 7:08:33 AM | Crash | batch

```
{{:badmatch,
  {:error, :push, 422,
   "{\"message\":\"Waiting on code owner review from dcoutts, erikd, Jimbo4350, mrBliss, and/or newhoggy.\",\"documentation_url\":\"https://docs.github.com/articles/about-protected-branches\"}"}},
 [
   {BorsNG.Worker.Batcher, :complete_batch, 3,
    [file: 'lib/worker/batcher.ex', line: 707]},
   {BorsNG.Worker.Batcher, :maybe_complete_batch, 1,
    [file: 'lib/worker/batcher.ex', line: 653]},
   {BorsNG.Worker.Batcher, :handle_cast, 2,
    [file: 'lib/worker/batcher.ex', line: 92]},
   {:gen_server, :try_dispatch, 4,
    [file: 'gen_server.erl', line: 637]},
   {:gen_server, :handle_msg, 6,
    [file: 'gen_server.erl', line: 711]},
   {:proc_lib, :init_p_do_apply, 3,
    [file: 'proc_lib.erl', line: 249]}
 ]}
```

**Comment by notriddle** [comment]
Do you have the `required_approvals` setting turned on in your bors.toml file?

**Comment by rvl** [comment]
Yes. `required_approvals = 1` ([bors.toml](https://github.com/input-output-hk/cardano-node/blob/0dc6efa467a0fdae7aba7c5bcd5c657e189c8f19/bors.toml#L7)). As I understand, bors-ng will accept approvals from anyone, even if they aren't codeowners. We do wish to keep the github codeowners branch protection enabled.

**Comment by notriddle** [comment]
Okay, so you need to enable the `use_codeowners = true` option.

**Comment by umamaistempo** [comment]
The report here is mainly about no error message being provided by the bors agent on the PR, @rvl ?

If so, I am going to take a look at the code to see ensure that the message is provided, what do you think @notriddle ?

**Comment by notriddle** [comment]
Sure, go ahead.

**Comment by rvl** [comment]
Ah so there is an option - sorry for the report I honestly did check the documentation last week, but missed `use_codeowners`.

Perhaps generic "github rejected push" error handling would be good, in case there are new ways for the API request to fail.

**Comment by tommilligan** [comment]
Hey @rvl, I believe this can be now closed as fixed by #1273. Please take a look and verify if it covers your use case, or provide more details if not.

bors now provides generic error handling if github rejects merging the tested commit into the parent branch.

**Comment by rvl** [comment]
Thanks @tommilligan!
